src/cpp/core/file_lock/FileLock.cpp

```diff
@@ -0,0 +1,257 @@
+   FileLock::s_timeoutInterval = boost::posix_time::seconds(timeoutInterval);
+   
+   // refresh rate
+   double refreshRate = getFieldPositive(settings, "refresh-rate", kDefaultRefreshRate);
+   FileLock::s_refreshRate = boost::posix_time::seconds(refreshRate);
+   
+   DEBUG_BLOCK("lock initialization")
+   {
+      std::cerr << "Type: "    << lockTypeToString(FileLock::s_defaultType) << std::endl;
+      std::cerr << "Timeout: " << FileLock::s_timeoutInterval.total_seconds() << std::endl;
+      std::cerr << "Refresh: " << FileLock::s_refreshRate.total_seconds() << std::endl;
+   }
+}
+
+// default values for static members
+FileLock::LockType FileLock::s_defaultType(FileLock::LOCKTYPE_LINKBASED);
+boost::posix_time::seconds FileLock::s_timeoutInterval(kDefaultTimeoutInterval);
+boost::posix_time::seconds FileLock::s_refreshRate(kDefaultRefreshRate);
+
+boost::shared_ptr<FileLock> FileLock::create(LockType type)
+{
+   verifyInitialized();
+   
+   switch (type)
+   {
+   case LOCKTYPE_ADVISORY:  return boost::shared_ptr<FileLock>(new AdvisoryFileLock());
+   case LOCKTYPE_LINKBASED: return boost::shared_ptr<FileLock>(new LinkBasedFileLock());
+   }
+   
+   // shouldn't be reached
+   return boost::shared_ptr<FileLock>(new AdvisoryFileLock());
+}
+
+boost::shared_ptr<FileLock> FileLock::createDefault()
+{
+   verifyInitialized();
+   return FileLock::create(s_defaultType);
+}
+
+void FileLock::refresh()
+{
+   verifyInitialized();
+   AdvisoryFileLock::refresh();
+   LinkBasedFileLock::refresh();
+}
+
+void FileLock::cleanUp()
+{
+   verifyInitialized();
+   AdvisoryFileLock::cleanUp();
+   LinkBasedFileLock::cleanUp();
+}
+
+namespace {
+
+void schedulePeriodicExecution(
+      const boost::system::error_code& ec,
+      boost::asio::deadline_timer& timer,
+      boost::posix_time::seconds interval,
+      boost::function<void()> callback)
+{
+   try
+   {
+      // bail on boost errors (these are very unexpected)
+      if (ec)
+      {
+         if (!core::isShutdownError(ec))
+            LOG_ERROR(core::Error(ec, ERROR_LOCATION));
+         return;
+      }
+      
+      // execute callback
+      callback();
+
+      // reschedule
+      boost::system::error_code errc;
+      timer.expires_at(timer.expires_at() + interval, errc);
+      if (errc)
+      {
+         LOG_ERROR(Error(errc, ERROR_LOCATION));
+         return;
+      }
+      
+      timer.async_wait(boost::bind(
+                          schedulePeriodicExecution,
+                          boost::asio::placeholders::error,
+                          boost::ref(timer),
+                          interval,
+                          callback));
+   }
+   catch (...)
+   {
+      // swallow errors
+   }
+}
+
+} // end anonymous namespace
+
+void FileLock::refreshPeriodically(boost::asio::io_service& service,
+                                   boost::posix_time::seconds interval)
+{
+   // protect against re-entrancy
```

Comment:
Not really 're-entrancy' but just want to ensure it's only ever called once.

It might be nice to have a macro that handles this, but I can't figure out what it should be named, or what it should look like...